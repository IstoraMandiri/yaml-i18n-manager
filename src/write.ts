import YAML from 'yaml';

import { resolve, dirname } from 'path';
import { promises } from 'fs';
import { replaceLocaleInPath } from './utils';

const { readFile, mkdir, writeFile } = promises;

function inflate(key: string, val: any, obj: any): any {
  const [frag, ...rest] = key.split('.');
  if (frag.startsWith('[') && frag.endsWith(']')) {
    const [_i, subKey] = frag.slice(1, -1).split(':');
    const i = parseInt(_i);
    const arr = obj || [];
    arr[i] = {
      ...inflate(rest.join('.'), val, arr[i]),
      key: subKey,
    };
    // Object.keys(obj)
    return arr;
  }
  if (rest.length === 0) {
    return {
      ...obj,
      [key]: val,
    };
  }
  if (key)
    return {
      ...obj,
      [frag]: inflate(rest.join('.'), val, obj[frag]),
    };
}

function getStructure(yamls: TranslatedYaml[]) {
  return yamls.reduce(
    (o: any, t) => ({ ...o, [t.path]: inflate(t.sig, t.update || t.value, o[t.path]) }),
    {},
  );
}

function merge(o: any, n: any): any {
  if (o === undefined) {
    return n;
  }
  if (n === undefined) {
    return o;
  }
  if (['string', 'boolean', 'number'].indexOf(typeof n) >= 0) {
    return n;
  }
  if (n instanceof Date) {
    return n;
  }
  // use `key` in arrays to match
  if (Array.isArray(n)) {
    if (!Array.isArray(o)) {
      return n;
    }
    return o.map((item) => {
      if (!item.key) {
        return item;
      }
      const match = n.find(({ key }) => key === item.key);
      return merge(item, match);
    });
  }
  // if it's an object...
  const res: { [key: string]: any } = { ...n };
  Object.keys(o).forEach((key) => {
    res[key] = merge(o[key], n[key]);
  });
  return res;
}

async function writeYaml(target: string, file: string, values: any) {
  // check if the file open
  const path = resolve(target, file);
  const dir = dirname(path);
  let data = values;
  // merge with existing values if they exist
  try {
    const fileContent = await readFile(path, 'utf8');
    const existingData = await YAML.parse(fileContent);
    data = merge(existingData, data);
    console.log({ existingData, data, values });
  } catch (e) {
    // do nothing
  }
  // write the file
  const newData = await YAML.stringify(data);
  await mkdir(dir, { recursive: true });
  await writeFile(path, newData);
}

export default async function write(yamls: TranslatedYaml[], target: string) {
  const structure = getStructure(yamls);
  await Promise.all(Object.keys(structure).map((key) => writeYaml(target, key, structure[key])));
}
