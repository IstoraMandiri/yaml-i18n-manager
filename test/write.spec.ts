import rimraf from 'rimraf';
import { resolve, relative } from 'path';
import { promises } from 'fs';
const { readdir, readFile } = promises;

import translatedYaml from './data/misc/translated.json';

import write from '../src/write';

async function scanFiles(root: string, dir?: string): Promise<any[]> {
  const target = dir || root;
  const dirEntries = await readdir(target, { withFileTypes: true });
  const files = await Promise.all(
    dirEntries.map(async (entry) => {
      const res = resolve(target, entry.name);
      return entry.isDirectory()
        ? scanFiles(root, res)
        : { fileName: relative(__dirname, res), content: await readFile(res, 'utf8') };
    }),
  );
  return Array.prototype.concat(...files);
}

const outDir = `./test/data/out-temp`;

describe('write', () => {
  beforeAll(() => rimraf.sync(outDir));
  afterAll(() => rimraf.sync(outDir));
  it('writes files correctly', async () => {
    await write(translatedYaml, outDir);
    expect(await scanFiles(outDir)).toMatchSnapshot();
  });
});
