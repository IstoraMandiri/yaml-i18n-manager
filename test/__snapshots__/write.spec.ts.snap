// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write merging merges correctly 1`] = `
Array [
  Object {
    "content": "\\"0\\":
  key: home
  title: ホーム
\\"1\\":
  key: news
  title: OLD
\\"2\\":
  key: special
  title: Don't delete me
\\"[0:home].title\\": ホーム
\\"[1:about].title\\": 約
\\"[2:news].title\\": ニュース
",
    "fileName": "test/data/merge-test/menu.ja.yaml",
  },
  Object {
    "content": "intro: このサブフォルダへようこそ
description: これは翻訳されない説明です
\\"items.[0:test].title\\": テスト中
\\"items.[1:hello].title\\": ハローワーク
items:
  - key: hello
    title: ハローワーク
",
    "fileName": "test/data/merge-test/sub/welcome.ja.yaml",
  },
  Object {
    "content": "title: 題名
description: これは説明です
new: 新しいキー
body: |
  ##これはいくつかの値下げです

  これはどのように？
",
    "fileName": "test/data/merge-test/ui.global.ja.yaml",
  },
]
`;

exports[`write writes freshly correctly 1`] = `
Array [
  Object {
    "content": "\\"[0:home].title\\": ホーム
\\"[1:about].title\\": 約
\\"[2:news].title\\": ニュース
",
    "fileName": "test/data/out-temp/menu.ja.yaml",
  },
  Object {
    "content": "intro: このサブフォルダへようこそ
description: これは翻訳されない説明です
\\"items.[0:test].title\\": テスト中
\\"items.[1:hello].title\\": ハローワーク
",
    "fileName": "test/data/out-temp/sub/welcome.ja.yaml",
  },
  Object {
    "content": "title: 題名
description: これは説明です
new: 新しいキー
body: |
  ##これはいくつかの値下げです

  これはどのように？
",
    "fileName": "test/data/out-temp/ui.global.ja.yaml",
  },
]
`;
